Oplog status metrics v2

timvaillancourt opened the pull request:
A continuation of #16

Moving to using 2 x .find() queries to find the min/max timestamp instead of a single, heavier .aggregate() query due to some issues under high oplog insert load where the aggregate sometimes fails. Also retrying once if unable to get the timestamps. This change resolved the "Failed to get oplog status." errors I noticed under heavy insert/update load.

dcu commented on collector/oplog_status.go:
```diff
@@ -57,18 +57,38 @@ func BsonMongoTimestampToUnix(timestamp bson.MongoTimestamp) float64 {
+
+	// retry once if there is an error
+	var tries int64 = 0
+	var head_result struct { Timestamp	bson.MongoTimestamp	`bson:"ts"` }
```
please use camel case to name vars

dcu commented on collector/oplog_status.go:
```diff
@@ -57,18 +57,38 @@ func BsonMongoTimestampToUnix(timestamp bson.MongoTimestamp) float64 {
+
+	// retry once if there is an error
+	tries = 0
+	var tail_result struct { Timestamp	bson.MongoTimestamp	`bson:"ts"` }
```
camel case instead of snake case please

dcu commented on collector/oplog_status.go:
```diff
@@ -57,18 +57,38 @@ func BsonMongoTimestampToUnix(timestamp bson.MongoTimestamp) float64 {
+	tries = 0
+	var tail_result struct { Timestamp	bson.MongoTimestamp	`bson:"ts"` }
+	for tries < 2 {
+		err = session.DB("local").C("oplog.rs").Find(nil).Sort("$natural").Limit(1).One(&tail_result)
```
is it possible to refactor this to not dup the code with the line 66?

timvaillancourt replied:
I agree @dcu, there's gotta be a better way. Let me get back to you with another commit

timvaillancourt commented:
Hey @dcu, I am happy with these changes now and have successfully tested.

Summary:
1. Moved to using a function that returns a single timestamp with a boolean for tail/head instead a struct of both tail/head, so that the code could be less duplicated.
2. Added sleep to retry, which was retrying immediately.
3. Refactored GetOplogStatus to error on the collection stats and timestamps separately, also made things a bit more readable.
4. Used camel case on variables.

dcu commented:
@timvaillancourt thanks, I merged this and also added you as collaborator

timvaillancourt commented:
Thanks @dcu!
